From bcf87d62617717356bad4d63122d324ee4580460 Mon Sep 17 00:00:00 2001
From: orbea <orbea@riseup.net>
Date: Sat, 14 May 2022 19:29:07 -0700
Subject: [PATCH] lib-index: Fix overlinking

When building dovecot with slibtool the build will fail with several
multiple definition errors.

This happens because the lib-index/Makefile.am links the tests with both
the libtool archives (.la) and libtool objects (.lo) that already exist
in the .la files.

GNU libtool does some magic to silently hide the issue while slibtool
explicitly does as its told which is where the bug is then exposed.

This can be solved by better sorting individual files into their own
convenience libraries to be used by both libindex.la and the tests as
needed. This also allows removing the many _DEPENDENCIES lines as they
are now redundant.

Gentoo Bug: https://bugs.gentoo.org/782631
---
 src/lib-index/Makefile.am | 81 ++++++++++++++++++++++-----------------
 1 file changed, 45 insertions(+), 36 deletions(-)

diff --git a/src/lib-index/Makefile.am b/src/lib-index/Makefile.am
index f203201bf7..ea2bcfbb72 100644
--- a/src/lib-index/Makefile.am
+++ b/src/lib-index/Makefile.am
@@ -1,10 +1,20 @@
-noinst_LTLIBRARIES = libindex.la
+index_libs = \
+	libmail-index-transaction-finish.la \
+	libmail-index-transaction-update.la \
+	libmail-index-util.la \
+	libmail-index-write.la \
+	libmail-transaction-log-append.la \
+	libmail-transaction-log-view.la
+
+noinst_LTLIBRARIES = libindex.la $(index_libs)
 
 AM_CPPFLAGS = \
 	-I$(top_srcdir)/src/lib \
 	-I$(top_srcdir)/src/lib-test \
 	-I$(top_srcdir)/src/lib-mail
 
+AM_LDFLAGS = -no-undefined
+
 libindex_la_SOURCES = \
 	mail-cache.c \
 	mail-cache-decisions.c \
@@ -24,26 +34,40 @@ libindex_la_SOURCES = \
         mail-index-modseq.c \
         mail-index-transaction.c \
         mail-index-transaction-export.c \
-        mail-index-transaction-finish.c \
         mail-index-transaction-sort-appends.c \
-        mail-index-transaction-update.c \
         mail-index-transaction-view.c \
         mail-index-strmap.c \
         mail-index-sync.c \
         mail-index-sync-ext.c \
         mail-index-sync-keywords.c \
         mail-index-sync-update.c \
-        mail-index-util.c \
         mail-index-view.c \
         mail-index-view-sync.c \
-        mail-index-write.c \
         mail-transaction-log.c \
-        mail-transaction-log-append.c \
         mail-transaction-log-file.c \
         mail-transaction-log-modseq.c \
-        mail-transaction-log-view.c \
         mailbox-log.c
 
+libindex_la_LIBADD = $(index_libs)
+
+libmail_index_transaction_finish_la_SOURCES = \
+	mail-index-transaction-finish.c
+
+libmail_index_transaction_update_la_SOURCES = \
+	mail-index-transaction-update.c
+
+libmail_index_util_la_SOURCES = \
+	mail-index-util.c
+
+libmail_index_write_la_SOURCES = \
+	mail-index-write.c
+
+libmail_transaction_log_append_la_SOURCES = \
+	mail-transaction-log-append.c
+
+libmail_transaction_log_view_la_SOURCES = \
+	mail-transaction-log-view.c
+
 headers = \
 	mail-cache.h \
 	mail-cache-private.h \
@@ -79,63 +103,48 @@ test_programs = \
 noinst_PROGRAMS = $(test_programs)
 
 test_libs = \
-	mail-index-util.lo \
+	libmail-index-util.la \
 	../lib-test/libtest.la \
 	../lib/liblib.la
 
-test_deps = $(noinst_LTLIBRARIES) $(test_libs)
-
 test_mail_cache_SOURCES = test-mail-cache-common.c test-mail-cache.c
-test_mail_cache_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_cache_DEPENDENCIES = $(test_deps)
+test_mail_cache_LDADD = libindex.la $(test_libs)
 
 test_mail_cache_fields_SOURCES = test-mail-cache-common.c test-mail-cache-fields.c
-test_mail_cache_fields_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_cache_fields_DEPENDENCIES = $(test_deps)
+test_mail_cache_fields_LDADD = libindex.la $(test_libs)
 
 test_mail_cache_purge_SOURCES = test-mail-cache-common.c test-mail-cache-purge.c
-test_mail_cache_purge_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_cache_purge_DEPENDENCIES = $(test_deps)
+test_mail_cache_purge_LDADD = libindex.la $(test_libs)
 
 test_mail_index_SOURCES = test-mail-index.c
-test_mail_index_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_index_DEPENDENCIES = $(test_deps)
+test_mail_index_LDADD = libindex.la $(test_libs)
 
 test_mail_index_map_SOURCES = test-mail-index-map.c
-test_mail_index_map_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_index_map_DEPENDENCIES = $(test_deps)
+test_mail_index_map_LDADD = libindex.la $(test_libs)
 
 test_mail_index_modseq_SOURCES = test-mail-index-modseq.c
-test_mail_index_modseq_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_index_modseq_DEPENDENCIES = $(test_deps)
+test_mail_index_modseq_LDADD = libindex.la $(test_libs)
 
 test_mail_index_sync_ext_SOURCES = test-mail-index-sync-ext.c
-test_mail_index_sync_ext_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_index_sync_ext_DEPENDENCIES = $(test_deps)
+test_mail_index_sync_ext_LDADD = libindex.la $(test_libs)
 
 test_mail_index_transaction_finish_SOURCES = test-mail-index-transaction-finish.c
-test_mail_index_transaction_finish_LDADD = mail-index-transaction-finish.lo $(test_libs)
-test_mail_index_transaction_finish_DEPENDENCIES = $(test_deps)
+test_mail_index_transaction_finish_LDADD = libmail-index-transaction-finish.la $(test_libs)
 
 test_mail_index_transaction_update_SOURCES = test-mail-index-transaction-update.c
-test_mail_index_transaction_update_LDADD = mail-index-transaction-update.lo $(test_libs)
-test_mail_index_transaction_update_DEPENDENCIES = $(test_deps)
+test_mail_index_transaction_update_LDADD = libmail-index-transaction-update.la $(test_libs)
 
 test_mail_index_write_SOURCES = test-mail-index-write.c
-test_mail_index_write_LDADD = mail-index-write.lo $(test_libs)
-test_mail_index_write_DEPENDENCIES = $(test_deps)
+test_mail_index_write_LDADD = libmail-index-write.la $(test_libs)
 
 test_mail_transaction_log_append_SOURCES = test-mail-transaction-log-append.c
-test_mail_transaction_log_append_LDADD = mail-transaction-log-append.lo $(test_libs)
-test_mail_transaction_log_append_DEPENDENCIES = $(test_deps)
+test_mail_transaction_log_append_LDADD = libmail-transaction-log-append.la $(test_libs)
 
 test_mail_transaction_log_file_SOURCES = test-mail-transaction-log-file.c
-test_mail_transaction_log_file_LDADD = $(noinst_LTLIBRARIES) $(test_libs)
-test_mail_transaction_log_file_DEPENDENCIES = $(test_deps)
+test_mail_transaction_log_file_LDADD = libindex.la $(test_libs)
 
 test_mail_transaction_log_view_SOURCES = test-mail-transaction-log-view.c
-test_mail_transaction_log_view_LDADD = mail-transaction-log-view.lo $(test_libs)
-test_mail_transaction_log_view_DEPENDENCIES = $(test_deps)
+test_mail_transaction_log_view_LDADD = libmail-transaction-log-view.la $(test_libs)
 
 check-local:
 	for bin in $(test_programs); do \
